cpp/unreal_plugins/SpEngine/Source/SpEngine/EngineService.h

```diff
@@ -0,0 +1,187 @@
+    ExecutingTick,
+    ExecutingPostTick,
+};
+
+template <CEntryPointBinder TEntryPointBinder>
+class EngineService {
+public:
+    EngineService() = delete;
+    EngineService(TEntryPointBinder* entry_point_binder)
+    {
+        SP_ASSERT(entry_point_binder);
+
+        entry_point_binder_ = entry_point_binder;
+
+        begin_frame_handle_ = FCoreDelegates::OnBeginFrame.AddRaw(this, &EngineService::beginFrameHandler);
+        end_frame_handle_ = FCoreDelegates::OnEndFrame.AddRaw(this, &EngineService::endFrameHandler);
+
+        frame_state_ = FrameState::Idle;
+
+        entry_point_binder_->bind("engine_service.begin_tick", [this]() -> void {
+            SP_ASSERT(frame_state_ == FrameState::Idle);
+
+            // reset promises and futures
+            frame_state_idle_promise_ = std::promise<void>();
+            frame_state_executing_pre_tick_promise_ = std::promise<void>();
+            frame_state_executing_post_tick_promise_ = std::promise<void>();
+
+            frame_state_idle_future_ = frame_state_idle_promise_.get_future();
+            frame_state_executing_pre_tick_future_ = frame_state_executing_pre_tick_promise_.get_future();
+            frame_state_executing_post_tick_future_ = frame_state_executing_post_tick_promise_.get_future();
+
+            // indicate that we want the game thread to advance the simulation, wait here until frame_state == FrameState::ExecutingPreTick
+            frame_state_ = FrameState::RequestPreTick;
+            frame_state_executing_pre_tick_future_.wait();
+
+            SP_ASSERT(frame_state_ == FrameState::ExecutingPreTick);
+        });
+
+        entry_point_binder_->bind("engine_service.tick", [this]() -> void {
+            SP_ASSERT(frame_state_ == FrameState::ExecutingPreTick);
+
+            // allow beginFrameHandler() to finish executing, wait here until frame_state == FrameState::ExecutingPostTick
+            work_queue_.reset();
+
+            frame_state_executing_post_tick_future_.wait();
+
+            SP_ASSERT(frame_state_ == FrameState::ExecutingPostTick);
+        });
+
+        entry_point_binder_->bind("engine_service.end_tick", [this]() -> void {
+            SP_ASSERT(frame_state_ == FrameState::ExecutingPostTick);
+
+            // allow endFrameHandler() to finish executing, wait here until frame_state == FrameState::Idle
+            work_queue_.reset();
+
+            frame_state_idle_future_.wait();
+
+            SP_ASSERT(frame_state_ == FrameState::Idle);
+        });
+
+        entry_point_binder_->bind("engine_service.get_byte_order", []() -> std::string {
+            uint32_t dummy = 0x01020304;
+            return (reinterpret_cast<char*>(&dummy)[3] == 1) ? "little" : "big";
+        });
+
+        entry_point_binder_->bind("engine_service.ping", []() -> std::string {
+            return "EngineService received a call to ping()...";
+        });
+
+        entry_point_binder_->bind("engine_service.request_close", []() -> void {
```

Comment:
Order the entry points as follows.

```
ping
request_close
begin_tick
tick
end_tick
get_byte_order
```

I'm trying to order the entry points from the most basic usage to the most advanced usage, so the most commonly needed information for a reader is at the top of the file. (Similar to how we would put `initialize()` and `terminate()` at the top of a header.) Obviously this ordering is somewhat subjective, but clearly the most basic possible usage of `EngineService` would be to call `ping` (it's the first thing anyone should do after connecting), and then call `request_close` immediately afterwards. The `tick` functions would be a slightly more advanced usage, and `get_byte_order` is even more advanced because you only need it if you're sending/receiving blocks of binary data.

In my most recent design sketch that demonstrates how to send/receive blocks of binary data to/from components, I have a dedicated service for handling all of the communication, managing shared memory, etc. I think `get_byte_order` should probably be in that service, but it's ok for this PR if we leave it here.